### PR TITLE
Fix tight loop in getFileInfo

### PR DESCRIPTION
**Pull request by ivankelly**
The argument to assertions is not evaluated if assertions is disabled,
which was messing up the refcounting for the fileinfo backing
cache. We ended up with dead fileinfo objects in the guava caches,
which triggered an infinite loop.

This patch fixes that by moving the tryRetain() out of the assert. It
also adds defensive code to getFileInfo to ensure that if a fileinfo
object is dead, that it doesn't exist any longer in the caches.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java [hunk 1]
```diff
@@ -220,7 +220,19 @@ CachedFileInfo getFileInfo(final Long ledger, final byte masterKey[]) throws IOE
...
                     fi = readFileInfoCache.get(ledger, loader);
                 }
-            } while (!fi.tryRetain());
+                if (!fi.tryRetain()) {
+                    // defensively ensure that dead fileinfo objects don't exist in the
+                    // cache. They shouldn't if refcounting is correct, but if someone
+                    // does a double release, the fileinfo will be cleaned up, while
+                    // remaining in the cache, which could cause a tight loop in this method.
+                    if (writeFileInfoCache.asMap().remove(ledger, fi)
+                        || readFileInfoCache.asMap().remove(ledger, fi)) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java]: if writeFileInfoCache#remove returns true we will miss to check readFileInfoCache#remove (the 'or' will not evaluate the second operand)

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java [hunk 2]
```diff
@@ -220,7 +220,19 @@ CachedFileInfo getFileInfo(final Long ledger, final byte masterKey[]) throws IOE
...
+                        LOG.error("Dead fileinfo({}) forced out of cache. Must have been double-released somewhere.",
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java]: Is there a way to automatically recover from this situation ? if not I would prefer to crash the bookie and not to run with a broken cache, what do you think @ivankelly  ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfoBackingCache.java [hunk 3]
```diff
@@ -52,7 +52,8 @@ CachedFileInfo loadFileInfo(long ledgerId, byte[] masterKey) throws IOException
                 // have been able to get a reference to it here.
                 // The caller of loadFileInfo owns the refence, and is
                 // responsible for calling the corresponding #release().
-                assert(fi.tryRetain());
+                boolean retained = fi.tryRetain();
+                assert(retained);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfoBackingCache.java]: if it is important that retained is true we should check it explicitly and crash, even if assertions are disabled

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java (hunk 1 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java]: Oops, fixed.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfoBackingCache.java (hunk 3 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfoBackingCache.java]: In this case, retained can only be false if someone screws up some code in this file. Noone else has a reference to fi at this point.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java (hunk 2 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/IndexPersistenceMgr.java]: automatic recover is removal from the caches. The fileinfo object is dead, so nothing can be done with it in any case.

**Comment by ivankelly** [comment]
@eolivelli regarding crashing, I don't have a very strong opinion, but lean against it. Any issue here is a programming error, but with the defensive code, it should be safe. I certainly don't want to put a System.exit in here, we'd need to add some interface stuff for crashing, and that's out of scope for this fix.

**Comment by eolivelli** [comment]
@ivankelly 

>  certainly don't want to put a System.exit in here, we'd need to add some interface stuff for crashing, and that's out of scope for this fix.

I agree at 100%

**Comment by sijie** [comment]
@ivankelly can you create a separate issue for cleaning up the usage of `assert`? we'd better not use `assert` in general. for tests, we need to use junit assertion, for other codes, we need to throw exceptions.  also if it is possible, add a checkstyle rule to enforce it.

**Comment by ivankelly** [comment]
@sijie created #995

**Comment by eolivelli** [comment]
merging now on master and branch-4.6, thank you @ivankelly

**Pull request closed without merging**